### DISCUSSION TITLE
Change user inside image to non-root

### DISCUSSION THREAD

**Discussion by NlVlN** [dani-garcia/vaultwarden#3295] [Ideas]
It is good practice to run containers as a non-root user, if possible.

The change will affect the ability to connect to priveleged ports (<1024) but traffic to ports 80/443 can be handled by a reverse proxy (recommended option for HTTPS anyway) or the ports can be published from the container to the host.

Dockerfile [USER](https://docs.docker.com/engine/reference/builder/#user) reference.

**Comment by BlackDex** [comment]
This will probably break a lot of instances of we would change it currently on the container side. I don't know if there is a good way for some form of effortlessly migration. Have not looked into this yet.

**Comment by NlVlN** [comment]
I assume that the change would have the biggest impact on users upgrading from version with root user to the one w/o as they’d have to adjust port mapping.
Also I’m not sure but it could cause an issue with mounted volume files permissions.

However it should not affect newly deployed instances if the documentation would reflect changes like docker commands, docker-compose configuration, etc.

**Reply by BlackDex** [reply]
That is what I'm afraid of. That means we can't change it without first warning about this in some way upfront.

Maybe there is a way, but not sure.

**Comment by jjlin** [comment]
This change would cause permission issues for the vast majority of users, so I don't think the benefits outweigh the amount of disruption it would create.

It's pretty easy to run as a non-root user for those who want to do that: https://github.com/dani-garcia/vaultwarden/wiki/Hardening-Guide#run-as-a-non-root-user

Binding to privileged ports in the container is also not really an issue these days, as discussed in the wiki link above.
